### PR TITLE
Allow pycbc_inference to apply multi-dimensional constraints from ini file

### PR DESCRIPTION
**Pull request by cmbiwer**
This PR:
  * Updates ``pycbc_inference``, ``option_utils``, and ``.set_p0`` functions to allow ``PriorEvaluator`` to use passed constraints
  * Fixes to emcee sampler trying ``write_state`` and writing samples.
  * Adds a ``constraint`` module with constraint definitions. The ``custom`` constraint takes a str and evaluates it.

Can use ini file like:
```
[constraint-1]
name = mtotal_lt ; pre-defined constraint
mtotal = 30

[constraint-2]
name = custom ; let's make things up
func = params["mass1"] + params["mass2"] < self.mtotal
mtotal = 30
```

### CODE REVIEW

**Comment by cmbiwer** [comment]
Depends on #1687.

**Comment by cmbiwer** [comment]
Updated so now can do:
```
[constraint-1]
name = cartesian_spin_space

[prior-xi1+xi2]
name = uniform_power_law
dim = 0.5
min-xi1 = 0.
max-xi1 = 1.
min-xi2 = 0.
max-xi2 = 1.

[prior-chi_eff+chi_a+phi_a+phi_s]
name = uniform
min-chi_eff = -1.
max-chi_eff = 1.
min-chi_a = -1.
max-chi_a = 1.
min-phi_a = 0.
max-phi_a = 6.283185307179586
min-phi_s = 0.
max-phi_s = 6.283185307179586
```

**Comment by cdcapano** [comment]
@cmbiwer Could you rebase this and update it when you get a chance? At the moment we can't do the same GW151226 analysis as was done by LALInference because of the cuts on chirp mass and mass ratio that were used; this would give us that ability.

I'm not sure you can apply arbitrary constraints to any distribution, as it changes the pdf. You can definitely apply constraints to a uniform distribution though; you just need to renormalize the pdf in that case. You can do that numerically; e.g., if you draw 1e6 samples and reject 90% of them, you just need to multiply the pdf by 10. It may be that you can do the same thing for any distribution, I'm not really certain. For now though, could you make these methods properties of the Uniform distribution?

**Comment by cmbiwer** [comment]
Hey Collin, sure. I really only intended to use this to apply a total mass cut on a uniform distribution of ``mass1``-``mass2``. I'll update it soon.

**Comment by cmbiwer** [comment]
I've rebase and update this PR.

Can test the normalization of ``logpdf`` with:
```
#! /usr/bin/env python

import numpy
from pycbc import distributions
from pycbc.distributions import constraints
from pycbc.inference import prior

def mass1_lt_10(params):
    """ Returns True if mass1 is below 10.
    """
    if isinstance(params.values()[0], float):
        return True if params["mass1"] < 10 else False
    else:
        return params["mass1"] < 10

# create prior with constraint
variable_args = ["mass1", "mass2"]
mass1 = (0, 100)
mass2 = (0, 100)
uniform_prior = distributions.Uniform(mass1=mass1, mass2=mass2)
prior_eval = prior.PriorEvaluator(["mass1", "mass2"], uniform_prior, constraints=[mass1_lt_10])

# print scale factor for PDF
print prior_eval._pdf_scale

# evaluate prior
print prior_eval(**{"mass1" : 2, "mass2" : 1})
print prior_eval(**{"mass1" : 20, "mass2" : 1})

# create prior without constraint but in same range as constraint
variable_args = ["mass1", "mass2"]
mass1 = (0, 10)
mass2 = (0, 100)
uniform_prior = distributions.Uniform(mass1=mass1, mass2=mass2)
prior_eval_2 = prior.PriorEvaluator(["mass1", "mass2"], uniform_prior)

# evaluate prior
print prior_eval_2(**{"mass1" : 2, "mass2" : 1})
print prior_eval_2(**{"mass1" : 20, "mass2" : 1})
```

And I test it with this ini file:
```
[variable_args]
; waveform parameters that will vary in MCMC
tc =
mass1 =
mass2 =
spin1_a =
spin1_azimuthal =
spin1_polar =
spin2_a =
spin2_azimuthal =
spin2_polar =
distance =
coa_phase =
inclination =
polarization =
ra =
dec =

[static_args]
; waveform parameters that will not change in MCMC
approximant = IMRPhenomPv2
f_lower = 19.0

[prior-tc]
; coalescence time prior
name = uniform
min-tc = 1126259461.9
max-tc = 1126259462.1

[prior-mass1]
name = uniform
min-mass1 = 3.
max-mass1 = 30.

[prior-mass2]
name = uniform
min-mass2 = 3.
max-mass2 = 30.

[prior-spin1_a]
name = uniform
min-spin1_a = 0.0
max-spin1_a = 0.99

[prior-spin1_polar+spin1_azimuthal]
name = uniform_solidangle
polar-angle = spin1_polar
azimuthal-angle = spin1_azimuthal

[prior-spin2_a]
name = uniform
min-spin2_a = 0.0
max-spin2_a = 0.99

[prior-spin2_polar+spin2_azimuthal]
name = uniform_solidangle
polar-angle = spin2_polar
azimuthal-angle = spin2_azimuthal

[prior-distance]
; distance prior
name = uniform_radius
min-distance = 10
max-distance = 2000

[prior-coa_phase]
; coalescence phase prior
name = uniform_angle

[prior-inclination]
; inclination prior
name = sin_angle

[prior-ra+dec]
; sky position prior
name = uniform_sky

[prior-polarization]
; polarization prior
name = uniform_angle

;
;   Sampling transforms
;
[sampling_parameters]
mass1, mass2 : mchirp, logitq
spin1_a : logitspin1a
spin2_a : logitspin2a

[sampling_transforms-mchirp+q]
; inputs mass1, mass2
; outputs mchirp, q
name = mass1_mass2_to_mchirp_q

[sampling_transforms-logitq]
name = logit
inputvar = q
outputvar = logitq
min-q = 1.
max-q = 10.

[sampling_transforms-logitspin1a]
name = logit
inputvar = spin1_a
outputvar = logitspin1a
min-spin1_a = 0.
max-spin1_a = 0.99

[sampling_transforms-logitspin2a]
name = logit
inputvar = spin2_a
outputvar = logitspin2a
min-spin2_a = 0.
max-spin2_a = 0.99

[constraint-1]
name = custom
func = params["mass1"] < 10
```

Where I added the constraint:
```
[constraint-1]
name = custom
func = params["mass1"] < 10
```

Note that you have access to the ``[variable_args]`` parameters when defining constraints.

File: pycbc/inference/prior.py [hunk 1]
```diff
@@ -103,6 +107,39 @@ def __init__(self, variable_args, *distributions, **kwargs):
...
+
+            # evaluate constraints
+            result = numpy.ones(len(samples.values()[0]), dtype=bool)
+            for constraint in self._constraints:
+                result = constraint(samples) & result
+
+            # set new scaling factor for prior to be
+            # the fraction of acceptances in random sampling of entire space
+            # times 100
+            self._pdf_scale = 100.0 * sum(result) / float(n_test_samples)
```
Comment (cdcapano) [inline pycbc/inference/prior.py]: Why the extra factor of 100?

File: pycbc/distributions/constraints.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    the constraint and False if they do not.
+    """
+    name = "custom"
+    required_parameters = []
+    def __init__(self, variable_args, **kwargs):
+
+        # set any given attributes and get transforms from variable_args
+        # to required parameters
+        for kwarg in kwargs.keys():
+            setattr(self, kwarg, kwargs[kwarg])
```
Comment (cdcapano) [inline pycbc/distributions/constraints.py]: I don't like having the func attribute set like this. It took me awhile to figure out how the constraint function below wasn't failing. Just make `func` a required argument. And are any other kwargs needed? If not, just delete this, since it makes it really hard to understand what this class may do.

File: pycbc/distributions/constraints.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        _, self.transforms = transforms.get_common_cbc_transforms(
+                                       self.required_parameters, variable_args)
+
+    def __call__(self, params):
+        """ Evaluates constraint.
+        """
+        params = transforms.apply_conversions(params, self.transforms) \
```
Comment (cdcapano) [inline pycbc/distributions/constraints.py]: Shouldn't that be `apply_transforms`?

File: pycbc/distributions/constraints.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+                     if self.transforms else params
+        return self._constraint(params)
+
+    def _constraint(self, params):
+        """ Evaluates constraint function.
+        """
+        return eval(self.func)
```
Comment (cdcapano) [inline pycbc/distributions/constraints.py]: I don't like having an eval here. We should limit how much these are used. I think it'd be better to use the eval that's in the guts of `FieldArray`. So, in `__call__`, before `apply_transforms`:
```
if isinstance(params, dict):
   params = FieldArray.from_kwargs(**params)
elif not isinstance(params, FieldArray):
   raise ValueError("params must be dict or FieldArray instance")
```
Then after the transforms are applied:
```
out = self._constraint(params)
if out.size == 1:
    out = out.item()
return out
```
Finally, change `_constraint` to:
```
return params[self.func]
```
This means that in the config file, you would only have to put:
```
[constraint-mchirp]
func = (mchirp >= 9.5) & (mchirp < 10.5)
```
To constrain the parameter in some bounds.

Although rename the `func` attribute to be more descriptive name. Maybe `constraint_arg`?

File: pycbc/distributions/constraints.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    name = "custom"
+    required_parameters = []
+    def __init__(self, variable_args, **kwargs):
+
+        # set any given attributes and get transforms from variable_args
+        # to required parameters
+        for kwarg in kwargs.keys():
+            setattr(self, kwarg, kwargs[kwarg])
+        _, self.transforms = transforms.get_common_cbc_transforms(
+                                       self.required_parameters, variable_args)
```
Comment (cdcapano) [inline pycbc/distributions/constraints.py]: Is it necessary to store the common cbc transforms and apply them below? This is only ever used in inference runs. I think it's reasonable to expect that contraints could only be applied to either sampling parameters or variable args. If some other derived coordinate is desired, then a transforms section could be added by the user to the config file to ensure that parameter is in the params dictionary that is passed to the prior evaluator class. If you remove having to create the transforms then you can get rid of the variable args input parameter, and just require that the constraints be initalized with a function string to apply. Also, it would save on re-applying all of these transforms at each iteration in the MCMC (although that's probably not a very expensive operation).

Comment (cmbiwer) [inline pycbc/distributions/constraints.py]: At the moment, I believe it is necessary to store the transforms. I'd prefer not as you say, its a waste and it gets confusing.

However, at the moment the ``LikelihoodEvaluator`` does not transform the coordinates. Correct me if I'm wrong though. Its done in the waveform generator. I think this should all be moved up to the ``LikelihoodEvaluator`` level and therefore both the prior and waveform generator can receive the transforms. I don't think changing that is in the scope of this PR though.

File: pycbc/inference/prior.py (hunk 1 above)
Comment (cmbiwer) [inline pycbc/inference/prior.py]: Opps ignore that. I've pushed the new commit that this was missing.

Comment (cmbiwer) [inline pycbc/inference/prior.py]: Err, I mean that removes this.

File: pycbc/distributions/constraints.py (hunk 2 above)
Comment (cmbiwer) [inline pycbc/distributions/constraints.py]: Ok, I've added ``func`` as an arg.

File: pycbc/distributions/constraints.py (hunk 4 above)
Comment (cmbiwer) [inline pycbc/distributions/constraints.py]: Alright thanks, I've changed it.

**Comment by cmbiwer** [comment]
Ok @cdcapano please take another look.
  * I've renamed ``func`` to ``constraint_arg`` and made it an arg to ``Constraint``.
  * Removed the 100.
  * Replaced ``eval`` with ``FieldArray``.

The above examples still work except in the ini file change the ``func`` to ``constraint_args`` and remove the ``params``, eg.:
```
[constraint-1]
name = custom
constraint_args = mass1 < 10
```

I left the transforms in ``PriorEvaluator``. There needs to be some rearranging of ``LikelihoodEvaluator`` and the waveform generators to make that all blend better with transformations. But this functionality is something I think we need to advance a couple projects / requests so I think we should include this now and update the classes later.
